compiler/rustc_metadata/src/locator.rs

```diff
@@ -912,21 +912,26 @@ impl CrateError {
                 );
                 let candidates = libraries
                     .iter()
-                    .filter_map(|(_, lib)| {
+                    .map(|(_, lib)| {
                         let crate_name = &lib.metadata.get_root().name().as_str();
-                        match (&lib.source.dylib, &lib.source.rlib) {
-                            (Some((pd, _)), Some((pr, _))) => Some(format!(
-                                "\ncrate `{}`: {}\n{:>padding$}",
-                                crate_name,
-                                pd.display(),
-                                pr.display(),
-                                padding = 8 + crate_name.len()
-                            )),
-                            (Some((p, _)), None) | (None, Some((p, _))) => {
-                                Some(format!("\ncrate `{}`: {}", crate_name, p.display()))
-                            }
-                            (None, None) => None,
+                        let mut paths = lib.source.paths();
+
+                        // This `unwrap()` should be okay because there has to be at least one
+                        // source file. `CrateSource`'s docs confirm that too.
+                        let mut s = format!(
+                            "\ncrate `{}`: {}",
+                            crate_name,
+                            paths.next().unwrap().display()
+                        );
+                        let padding = 8 + crate_name.len();
+                        for path in paths {
+                            s.push_str(&format!(
+                                "\n{:>padding$}",
+                                path.display(),
+                                padding = padding
+                            ));
```

Comment:
I switched to `push_str(&format!(...))` from `write!` because `std::io::Write` is already in scope, so importing `std::fmt::Write` causes a name conflict. This code is in the error path, so the performance difference shouldn't matter.

Reply:
You can do `use io::Write as _` if you like.